plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+applyEvaluateBuiltinName heap env name =
+    runIdentity . runEvaluate (const $ Identity . evalL heap env) . runQuoteT . applyBuiltinName name
+
+evalL :: Heap -> Environment -> Plain Term -> EvaluationResult
+evalL heap env term = translateResult $ computeL [] heap (Closure term env)
+-- I'm not sure if this is doing quite the right thing.  The current
+-- strategy of the dynamic interface is that it when you're evaluating
+-- a builtin, you also pass the current PLC evaluator along with its
+-- state so that the builtin can evaluate terms properly (for example,
+-- it may need to look up a variable in an environment).  It's
+-- possible that this could update the heap, but in this case we're
+-- not getting the new heap back.  I think this is harmless, but
+-- perhaps inefficient (we may have to re-evaluate something that
+-- we've already evaluated).  We could solve this by putting the heap
+-- in a suitable monad, but let's wait until the interface has
+-- stabilised first.  I think the version above is OK for simple
+-- builtins for the time being.
+
+
+-- | Instantiate a term with a type and proceed.
+-- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
+-- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
+-- apply the term to the type via 'TyInst'.
+instantiateEvaluate :: EvaluationContext -> Heap -> Type TyName () -> Closure -> LMachineResult
+instantiateEvaluate ctx heap ty (Closure fun env) =
+    case fun of
+      TyAbs _ _ _ body ->
+          computeL ctx heap (Closure body env)
+      _ -> if isJust $ termAsPrimIterApp fun
+           then returnL ctx heap $ Closure (TyInst () fun ty) env
+           else throw $ LMachineException "Attempting to instantiate invalid term" fun
+
+-- | Evaluate a term using the L machine. This internal version
+-- returns a result containing the final heap and environment, which
+-- it mightbe useful to know (for example, if we're testing we might
+-- want to know how big the heap is.  Also, if we want to return a
+-- value to whatever has invoked the machine we may want to fully
+-- expand it, and you'd need the environment and heap to do that.
+internalEvaluateL :: Term TyName Name () -> LMachineResult
+internalEvaluateL t = computeL [] emptyHeap (Closure t emptyEnvironment)
+
+
+-- | Convert an L machine result into the standard result type for communication with the outside world.
+translateResult :: LMachineResult -> EvaluationResult
+translateResult r = case r of
+      Success (Closure t _) (Heap _ _) -> EvaluationSuccess t
+      Failure                          -> EvaluationFailure
+
+-- | Evaluate a term using the L machine. May throw an 'LMachineException'.
+evaluateL :: Term TyName Name () -> EvaluationResult
+evaluateL term = translateResult $  internalEvaluateL term
+
+-- | Run a program using the L machine. May throw a 'MachineException'.
+-- We're not using the dynamic names at the moment, but we'll require them eventually
```

Comment:
Oh, I thought you were using them. We can abstract this part out of the CEK machine I guess, because there likely will be duplication.